### ISSUE TITLE
MultiTrackInterpolantAutomatonBuilder may create broken automata

### ISSUE THREAD

**Issue by danieldietsch** [ultimate-pa/ultimate#141]
When Taipan creates an interpolant automaton with the ``MultiTrackInterpolantAutomatonBuilder `` (see [TaipanRefinementStrategy.java#L308](https://github.com/ultimate-pa/ultimate/blob/0564707841f6f04d5c884dc2443ddd4e7cbcb106/trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/tracehandling/TaipanRefinementStrategy.java#L308)), the automaton triggers an assertion during ``RemoveUnreachable``:
````
java.lang.AssertionError
	at de.uni_freiburg.informatik.ultimate.automata.nestedword.reachablestates.NestedWordAutomatonReachableStates.checkTransitionsReturnedConsistent(NestedWordAutomatonReachableStates.java:830)
	at de.uni_freiburg.informatik.ultimate.automata.nestedword.reachablestates.NestedWordAutomatonReachableStates$ReachableStatesComputation.<init>(NestedWordAutomatonReachableStates.java:1001)
	at de.uni_freiburg.informatik.ultimate.automata.nestedword.reachablestates.NestedWordAutomatonReachableStates.<init>(NestedWordAutomatonReachableStates.java:180)
	at de.uni_freiburg.informatik.ultimate.automata.nestedword.operations.RemoveUnreachable.<init>(RemoveUnreachable.java:68)
````

Reproduce with 
 * I:programs\abstractInterpretation\regression\all\proc-implies-statesplit.bpl  
 * S:settings\ai\svcomp-Reach-32bit-Taipan_Default_Debug.epf  
 * T:toolchains\AutomizerBpl.xml

**Comment by Heizmann** [comment]
Problem:
The equals comparison is not symmetric. The automaton has to deal with two objects that look quite similar and whose serial number is #15. One is a BasicPredicate, the other is an AbsIntPredicate.
If we compare both using the equals method of BasicPredicate, the answer is false.
If we compare both using the equals method of the AbsIntPredicate, the answer is true.

The automata library requires that the equals comparison is symmetric.
I have no idea how I can give a better error message.

**Comment by danieldietsch** [comment]
You are right, this seems to be the problem. 
I made the mistake and did not express pre and prostconditions as AbsIntPredicate; I solved this with 	bf415b6, but now I fail because [``BasicAbstractInterpolantAutomaton`` takes its true-state directly from the predicate unifier](https://github.com/ultimate-pa/ultimate/blob/afada407ee5c665436f3a1fe5cc55d0cf1f73986/trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/interpolantautomata/transitionappender/BasicAbstractInterpolantAutomaton.java#L60).

The ``PredicateUnifier`` creates true and false states by calling its predicatefactory and creating them via ``true`` and ```false`` terms. 

As I see it, we can either remove AutomataOperations calls to things like PredicateUnifier and PredicateFactory, or decouple the PredicateUnifier from the PredicateFactory (they should not communicate via their internals, terms), or write a completely new combination of PredicateUnifier and PredicateFactory for AbsIntPredicates. 

A quick fix is replacing the original equals() code of BasePredicate 
````
	public boolean equals(final Object obj) {
		if (this == obj) {
			return true;
		}
		if (obj == null) {
			return false;
		}
		if (getClass() != obj.getClass()) {
			return false;
		}
		final BasicPredicate other = (BasicPredicate) obj;
		return mSerialNumber == other.mSerialNumber;
	}
````
with 
````
	public boolean equals(final Object obj) {
		if (this == obj) {
			return true;
		}
		if (obj == null) {
			return false;
		}
		if (!(obj instanceof IPredicate)) {
			return false;
		}
		final IPredicate other = (IPredicate) obj;
		return hashCode() == other.hashCode();
	}
````
